Add note re: create a replicator for every resource you want to replicate

hillaryfraley opened the pull request:
## Description
- Adds note and critical success factor re: creating a replicator for every resource in a namespace in EtcdReplicators reference
- Adds note re: creating a replicator for every resource in a namespace in federation guide

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2222

hillaryfraley committed:
Add note to explain one replicator per resource to replicate

cwjohnston commented on content/sensu-go/5.18/guides/use-federation.md:
```diff
@@ -170,6 +170,12 @@ Etcd replicators use the [etcd make-mirror utility][12] for one-way replication
 To get started, configure one etcd replicator per cluster for each of those RBAC policy types, across all namespaces, for each backend in the federation.
+
+{{% notice note %}}
+**NOTE**: Create a replicator for each resource you want to replicate within a namespace.
```
```suggestion
**NOTE**: Create a replicator for each resource type you want to replicate within a namespace.
```

hillaryfraley replied:
Yes, I think you're right about that. fixing now...

cwjohnston replied:
"within a namespace" confuses me here.

cwjohnston commented on content/sensu-go/5.18/reference/etcdreplicators.md:
```diff
@@ -39,7 +39,12 @@ The replicator process watches the keyspace of the resource to be replicated and
+This is harmless, but you should consider it when designing a replicated system.
+
+{{% notice note %}}
+**NOTE**: Create a replicator for each resource you want to replicate within a namespace.
```
```suggestion
**NOTE**: Create a replicator for each resource type you want to replicate within a namespace.
```

cwjohnston commented on content/sensu-go/5.18/reference/etcdreplicators.md:
```diff
@@ -380,6 +385,10 @@ Before you implement etcd replicators, review these details &mdash; they are cri
 
 - Never use insecure mode without TLS mutual authentication outside of a testbed.
 
+**Create a replicator for each resource you want to replicate within a namespace.**
```
```suggestion
**Create a replicator for each resource type you want to replicate within a namespace.**
```

cwjohnston reviewed (changes requested):
If I understand correctly, we're talking about resource *types*, not individual named resources. That distinction wasn't clear to me here. IMO it also needs to be clarified in the EtcdReplicator specification.

hillaryfraley committed:
Change "resource" to "resource type"

Co-Authored-By: Cameron Johnston <cameron@sensu.io>

cwjohnston commented on content/sensu-go/5.18/guides/use-federation.md:
```diff
@@ -170,6 +170,12 @@ Etcd replicators use the [etcd make-mirror utility][12] for one-way replication
+
+{{% notice note %}}
+**NOTE**: Create a replicator for each resource type you want to replicate within a namespace.
+Creating a replicator for a particular namespace will **not** replicate all resources within that namespace.
```
```suggestion
**NOTE**: Create a replicator for each resource type you want to replicate. 
Replicating `Namespace` resources will **not** replicate the resources that belong to those namespaces.
```

cwjohnston commented on content/sensu-go/5.18/reference/etcdreplicators.md:
```diff
@@ -380,6 +385,10 @@ Before you implement etcd replicators, review these details &mdash; they are cri
 
+**Create a replicator for each resource type you want to replicate within a namespace.**
+
+- If you only create a replicator for a namespace, it will **not** replicate all resources within that namespace.
```
```suggestion
**Create a replicator for each resource type you want to replicate**

- Replicating `Namespace` resources will **not** replicate the resources that belong to those namespaces.
```

cwjohnston commented on content/sensu-go/5.18/reference/etcdreplicators.md:
```diff
@@ -39,7 +39,12 @@ The replicator process watches the keyspace of the resource to be replicated and
+
+{{% notice note %}}
+**NOTE**: Create a replicator for each resource type you want to replicate within a namespace.
+Creating a replicator for a particular namespace will **not** replicate all resources within that namespace.
```
```suggestion
**NOTE**: Create a replicator for each resource type you want to replicate.
Replicating `Namespace` resources will **not** replicate the resources that belong to those namespaces.
```

cwjohnston reviewed (changes requested):
Suggested to rephrasing for clarity

hillaryfraley committed:
Apply suggestions from code review

Co-Authored-By: Cameron Johnston <cameron@sensu.io>

hillaryfraley committed:
Merge branch 'master' into replicators-note

hillaryfraley committed:
Add updates from Cameron's review and propagate in 5.16 and 5.17

hillaryfraley committed:
Merge branch 'master' into replicators-note

nikkictl reviewed (commented):
TBH it seems a little redundant to have the same message in every single API route. I think having it once in the API docs, once in the reference, and once in the guide is good enough.

hillaryfraley committed:
Merge branch 'master' into replicators-note

hillaryfraley commented:
@nikkixdev I think it is listed only once in every doc except the Federation reference, where it's listed once in the body of the doc and then later as a critical success factor. I moved a different note about namespaces into the description tables in the API doc in this PR too (scope creep -- I'm sorry).

nikkictl commented on content/sensu-go/5.16/api/federation.md:
```diff
@@ -68,7 +66,7 @@ http://127.0.0.1:8080/api/enterprise/federation/v1/etcd-replicators \
 /etcd-replicators (GET)  | 
 ---------------|------
-description    | Returns the list of replicators.
+description    | Returns the list of replicators. _**NOTE**: If you did not specify a [namespace][2] when you created a replicator, the response will not include a `namespace` key-value pair._
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.16/api/federation.md:
```diff
@@ -98,7 +96,8 @@ output         | {{< highlight shell >}}
 The `/etcd-replicators` API endpoint provides HTTP POST access to create replicators.
 
-_**NOTE**: If you do not specify a [namespace][2] when you create a replicator, all namespaces for the given resource are replicated._
+_**NOTE**: Create a replicator for each resource type you want to replicate. 
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.16/api/federation.md:
```diff
@@ -134,7 +133,7 @@ HTTP/1.1 200 OK
 /etcd-replicators (POST) | 
 ----------------|------
-description     | Creates a new replicator (if none exists).
+description     | Creates a new replicator (if none exists). _**NOTE**: If you do not specify a [namespace][2] when you create a replicator, all namespaces for the given resource are replicated._
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.16/api/federation.md:
```diff
@@ -196,7 +193,7 @@ http://127.0.0.1:8080/api/enterprise/federation/v1/etcd-replicators/my_replicato
 /etcd-replicators/:etcd-replicator (GET) | 
 ---------------------|------
-description          | Returns the specified replicator.
+description          | Returns the specified replicator. _**NOTE**: If you did not specify a [namespace][2] when you created the replicator, the response will not include a `namespace` key-value pair._
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.17/api/federation.md:
```diff
@@ -68,7 +66,7 @@ http://127.0.0.1:8080/api/enterprise/federation/v1/etcd-replicators \
 /etcd-replicators (GET)  | 
 ---------------|------
-description    | Returns the list of replicators.
+description    | Returns the list of replicators. _**NOTE**: If you did not specify a [namespace][2] when you created a replicator, the response will not include a `namespace` key-value pair._
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.17/api/federation.md:
```diff
@@ -98,7 +96,8 @@ output         | {{< highlight shell >}}
 The `/etcd-replicators` API endpoint provides HTTP POST access to create replicators.
 
-_**NOTE**: If you do not specify a [namespace][2] when you create a replicator, all namespaces for the given resource are replicated._
+_**NOTE**: Create a replicator for each resource type you want to replicate. 
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.17/api/federation.md:
```diff
@@ -134,7 +133,7 @@ HTTP/1.1 200 OK
 /etcd-replicators (POST) | 
 ----------------|------
-description     | Creates a new replicator (if none exists).
+description     | Creates a new replicator (if none exists). _**NOTE**: If you do not specify a [namespace][2] when you create a replicator, all namespaces for the given resource are replicated._
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.17/api/federation.md:
```diff
@@ -196,7 +193,7 @@ http://127.0.0.1:8080/api/enterprise/federation/v1/etcd-replicators/my_replicato
 /etcd-replicators/:etcd-replicator (GET) | 
 ---------------------|------
-description          | Returns the specified replicator.
+description          | Returns the specified replicator. _**NOTE**: If you did not specify a [namespace][2] when you created the replicator, the response will not include a `namespace` key-value pair._
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.18/api/federation.md:
```diff
@@ -72,7 +68,9 @@ http://127.0.0.1:8080/api/enterprise/federation/v1/etcd-replicators \
 ---------------|------
-description    | Returns the list of replicators.
+description    | Returns the list of replicators. {{% notice note %}}
+**NOTE**: If you did not specify a [namespace](../../reference/etcdreplicators#namespace-attribute) when you created a replicator, the response will not include a `namespace` key-value pair.
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.18/api/federation.md:
```diff
@@ -103,7 +101,8 @@ output         | {{< highlight shell >}}
 
 {{% notice note %}}
-**NOTE**: If you do not specify a [namespace](../../reference/etcdreplicators#namespace-attribute) when you create a replicator, all namespaces for the given resource are replicated.
+**NOTE**: Create a replicator for each resource type you want to replicate. 
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.18/api/federation.md:
```diff
@@ -140,7 +139,9 @@ HTTP/1.1 200 OK
 ----------------|------
-description     | Creates a new replicator (if none exists).
+description     | Creates a new replicator (if none exists). {{% notice note %}}
+**NOTE**: If you do not specify a [namespace](../../reference/etcdreplicators#namespace-attribute) when you create a replicator, all namespaces for the given resource are replicated.
```
This note can be moved to the top to avoid duplication.

nikkictl commented on content/sensu-go/5.18/api/federation.md:
```diff
@@ -204,7 +201,9 @@ http://127.0.0.1:8080/api/enterprise/federation/v1/etcd-replicators/my_replicato
 ---------------------|------
-description          | Returns the specified replicator.
+description          | Returns the specified replicator. {{% notice note %}}
+**NOTE**: If you did not specify a [namespace](../../reference/etcdreplicators#namespace-attribute) when you created the replicator, the response will not include a `namespace` key-value pair.
```
This note can be moved to the top to avoid duplication.

nikkictl reviewed (commented):
See duplicate entries in the API docs. I don't think it needs to be there for every API endpoint since all endpoints will behave the same according to the note.

hillaryfraley commented:
@nikkixdev I see what you are saying, but I think the duplication is warranted. The note `If you did not specify a namespace when you created a replicator, the response will not include a `namespace` key-value pair.` makes more sense with the individual endpoints that have a response that lists any key-value pairs. It is easier to understand what that means if it's with one of the endpoints that has a response with key-value pairs. This is the note I mentioned that is kind of out of scope for this PR. The `Create a replicator for each resource type you want to replicate.` note is only included once in each document, except for in the etcd replicators reference as I mentioned.

Moving the notes to the top means combining them into one longer note, which I think could be confusing because we are explaining two different concepts that both mention namespaces but aren't the same. I also don't want to assume that everyone will read the doc from beginning to end. They may click straight to one endpoint without reading the top of the doc.

nikkictl reviewed (commented):
I'm sorry I'm not sure I understand the "out of scope" information that's included then. Maybe someone a little more informed on etcd replicators can better review this, perhaps @echlebek.

echlebek commented:
@nikkixdev @hillaryfraley I don't want to prescribe how the docs should look here, but I will note that every type of resource in the system that works at a global level, that is, doesn't have a namespace, will omit the namespace key in its response.

Whether or not that warrants a note in every circumstance, I'm not sure.

hillaryfraley commented:
I think I have definitely overcomplicated this one! This PR is intended to address https://trello.com/c/DRqEYZdg (I created the issue https://github.com/sensu/sensu-docs/issues/2222 after the 5.18 retro).

edit: The note that is duplicated already existed in the API doc. It was just listed with the example rather than in the description table. I moved it so I could add the `Create a replicator for each resource type you want to replicate` note in spot that seemed to make more sense.

hillaryfraley commented:
After sleeping on this, I think we just need to make sure this note is correct to resolve https://github.com/sensu/sensu-docs/issues/2222:

`Create a replicator for each resource type you want to replicate. Replicating `namespace` resources will **not** replicate the resources that belong to those namespaces.`

nikkictl commented:
@hillaryfraley `Create a replicator for each resource type you want to replicate. Replicating namespace resources will **not** replicate the resources that belong to those namespaces.` is sufficient to close https://github.com/sensu/sensu-docs/issues/2222.

`**NOTE**: If you did not specify a [namespace][2] when you created a replicator, the response will not include a namespace key-value pair.` however does not make sense to me so I can't approve this until I understand the justification for these additional changes.

hillaryfraley committed:
Merge branch 'master' into replicators-note

hillaryfraley commented:
The note `**NOTE**: If you did not specify a [namespace][2] when you created a replicator, the response will not include a namespace key-value pair.` is in the existing doc word-for-word (see https://docs.sensu.io/sensu-go/latest/api/federation/#etcd-replicators-get-example and ~https://docs.sensu.io/sensu-go/latest/api/federation/#etcd-replicators-post~ https://docs.sensu.io/sensu-go/latest/api/federation/#etcd-replicatorsetcd-replicator-get-example).

In this PR, I just moved it from the "Example" subsection to the "API specification" table for the same endpoints, without making any changes to the text. I moved it because I thought it made better sense in the table and because I wanted to add the note that closes #2222 in the [POST /etcd-replicators](https://docs.sensu.io/sensu-go/latest/api/federation/#etcd-replicators-post) section without combining the two notes (edit: which required moving another note).

It sounds like I need to create a separate issue to rewrite `**NOTE**: If you did not specify a [namespace][2] when you created a replicator, the response will not include a namespace key-value pair.` so that it makes sense.

nikkictl reviewed (approved):
Like mentioned in https://github.com/sensu/sensu-docs/pull/2225#issuecomment-593615508, all resources at the global level will omit the namespace key-value pair. IMO it doesn't make sense to make that distinction in every specific endpoint and specifically for replicators. However I've shared that, and regardless if it happens in this PR or a separate one, the change will likely get merged anyway.

hillaryfraley commented:
It may be better suited for the API overview. I don't know why that note is only in the federation API doc, but I should figure out whether there is a reason and make sure it ends up in the right place. Thank you for bringing it up.
